src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1637 @@
+      return !(*this == rhs);
+    }
+
+    omap_inner_key_t get_node_key() const {
+      omap_inner_key_le_t kint = node->get_node_key_ptr()[index];
+      return omap_inner_key_t(kint);
+    }
+
+    char *get_node_val_ptr() {
+      auto tail = node->buf + BLOCK_SIZE;
+      if (*this == node->iter_end())
+        return tail;
+      else {
+        return tail - static_cast<uint32_t>(get_node_key().key_off);
+      }
+    }
+
+    const char *get_node_val_ptr() const {
+      auto tail = node->buf + BLOCK_SIZE;
+      if ( *this == node->iter_end())
+        return tail;
+      else {
+        return tail - static_cast<uint32_t>(get_node_key().key_off);
+      }
+    }
+
+    void set_node_val(std::string_view val) {
+      static_assert(!is_const);
+      std::strcpy((char*)get_node_val_ptr(), std::string(val).c_str()); //copy char* to char* include "\0"
+    }
+
+    std::string get_node_val(){
+     std::string s(get_node_val_ptr());
+     return s;
+    }
+    std::string get_node_val() const{
+      std::string s(get_node_val_ptr());
+      return s;
+    }
+
+    bool contains(std::string_view key) const {
+      auto next = *this + 1;
+      if (next == node->iter_end())
+        return get_node_val() <= key;
+
+      return (get_node_val() <= key) && (next->get_node_val() > key);
+    }
+
+    uint16_t get_index() const {
+      return index;
+    }
+
+  private:
+    int get_right_offset() const {
+      return get_node_key().key_off;
+    }
+
+    int get_right_offset_end() const {
+      if (index == 0)
+        return 0;
+      else
+        return (*this - 1)->get_right_offset();
+    }
+
+    char *get_right_ptr() {
+      return node->buf + BLOCK_SIZE - get_right_offset();
+    }
+
+    const char *get_right_ptr() const {
+      static_assert(!is_const);
+      return node->buf + BLOCK_SIZE - get_right_offset();
+    }
+
+    char *get_right_ptr_end() {
+      if (index == 0)
+        return node->buf + BLOCK_SIZE;
+      else
+        return (*this - 1)->get_right_ptr();
+    }
+
+    const char *get_right_ptr_end() const {
+      if (index == 0)
+        return node->buf + BLOCK_SIZE;
+      else
+        return (*this - 1)->get_right_ptr();
+    }
+
+    void update_offset(int offset) {
+      auto key = get_node_key();
+      assert(offset + key.key_off >= 0);
+      key.key_off += offset;
+      set_node_key(key);
+    }
+
+    void set_node_key(omap_inner_key_t _lb) const {
+      static_assert(!is_const);
+      omap_inner_key_le_t lb;
+      lb = _lb;
+      node->get_node_key_ptr()[index] = lb;
+    }
+
+    typename crimson::common::maybe_const_t<char, is_const>::type get_node_key_ptr() const {
+      return reinterpret_cast<
+        typename crimson::common::maybe_const_t<char, is_const>::type>(
+	        node->get_node_key_ptr() + index);
+    }
+
+  };
+  using const_iterator = iter_t<true>;
+  using iterator = iter_t<false>;
+
+  struct delta_inner_t {
+    enum class op_t : uint8_t {
```

Comment:
Is the size very important? if not, might consider uint_fast8_t.
At least on ARMs - it makes a difference